x-pack/metricbeat/module/sql/query/dsn.go

```diff
@@ -0,0 +1,42 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package query
+
+import (
+	"net/url"
+
+	"github.com/go-sql-driver/mysql"
+
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// ParseDSN tries to parse the host
+func ParseDSN(mod mb.Module, host string) (mb.HostData, error) {
+	// TODO: Add support for `username` and `password` as module options
+
+	sanitized := sanitize(host)
+
+	return mb.HostData{
+		URI:          host,
+		SanitizedURI: sanitized,
+		Host:         sanitized,
+	}, nil
+}
+
+func sanitize(host string) string {
+	// Host is a standard URL
+	if url, err := url.Parse(host); err == nil && len(url.Host) > 0 {
+		return url.Host
+	}
+
+	// Host is a MySQL DSN
+	if config, err := mysql.ParseDSN(host); err == nil {
+		return config.Addr
+	}
+
+	// TODO: Add support for PostgreSQL connection strings and other formats
```

Comment:
would this mean that we need to add support for most existing db engines? This wasn't the case with the previous `datasource` field, right?

Reply:
This is only needed to have a proper value in `source.address`. We cannot use the DSN in `hosts` directly (or the previous `datasource`), because it can contain credentials.

With the `datasource` option the value in `hosts` was being used for `source.address`, but not to make the real connections, what was inconsistent.

I.e. with this configuration:
```
- module: sql
  hosts: ['example.com:3306']
  datasource: "mysql://user@password:db01.local:3306/"
  ...
```
The module was connecting to `db01.local`, but the value in `source.address` would be `example.com:3306`.

With the changes in this PR, the configuration has to be like this:
```
- module: sql
  hosts: ['mysql://user@password:db01.local:3306/']
  ...
```
In this case we need to parse this DSN to get the real address and don't expose passwords. Different drivers accept different DSN formats. This is what we may need multiple parsers here. 

This doesn't affect data collection, data collection will work with any driver and DSN format even if we cannot parse it here.